### PR TITLE
Added lcollapse and rcollapse operator

### PR DESCRIPTION
**Pull request by luitjens**

### CODE REVIEW

File: include/matx_tensor_ops.h [hunk 1]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+    index_t size_;  // size of collapsed dim
+
+  public:
+    using matxop = bool;
+    using scalar_type = typename T1::scalar_type;
+
+    __MATX_INLINE__ LCollapseOp(const T1 &op) : op_(op)
+    {
+      static_assert(DIM < T1::Rank(),  "Collapse DIM must be less than Rank() of operator");
+      static_assert(DIM > 0, "Collapse DIM must have a magnitude greather than 0");
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1111]: greater

File: include/matx_tensor_ops.h [hunk 2]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+      std::array<index_t, Rank()> in{indices...};  // index coming in
+      std::array<index_t, T1::Rank()> out;         // index going out
+
+      #pragma unroll
+      for(int i = 1; i < Rank(); i++) {
+        // copy all but first input index into out array
+        out[DIM+i] = in[i];
+      }
+       
+      // expand firs input index into DIM indices
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1137]: first

File: include/matx_tensor_ops.h [hunk 3]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+ * lcollapse operator
+ *
+ * The lcollapse operator takes a tensor and collapses the left most dimensions into a single dimension.
+ *
+ * @tparam DIM
+ *   The number of dimensions to collapse
+ * @tparam T1
+ *   Operator type
+ *
+ * @returns
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1173]: Missing `a` param

File: include/matx_tensor_ops.h [hunk 4]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+
+  public:
+    using matxop = bool;
+    using scalar_type = typename T1::scalar_type;
+    using matxlvalue = bool;
+
+    __MATX_INLINE__ RCollapseOp(const T1 &op) : op_(op)
+    {
+      static_assert(DIM < T1::Rank(),  "Collapse DIM must be less than Rank() of operator");
+      static_assert(DIM > 0, "Collapse DIM must have a magnitude greather than 0");
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1198]: greater

File: include/matx_tensor_ops.h [hunk 5]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+      static_assert(T1::Rank() > 1, "Collapse must be called on operators with rank > 1");
+
+      // comptue size of collapsed dimension
+      size_ = 1;
+      
+      // Collapse right-most dims
+      #pragma unroll
+      for(int i = 0 ; i <= DIM; i++) {
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1206]: Could instead do `i = Rank(); i >= Rank() - DIM; i--` or something

File: include/matx_tensor_ops.h [hunk 6]
```diff
@@ -1092,7 +1092,179 @@ auto __MATX_INLINE__ reverse(Op t)
...
+  }
+/**
+ * rcollapse operator
+ *
+ * The rcollapse operator takes a tensor and collapses the right most dimensions into a single dimension.
+ *
+ * @tparam DIM
+ *   The number of dimensions to collapse
+ * @tparam T1
+ *   Operator type
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:1258]: Missing `a` parameter

File: include/matx_tensor_ops.h (hunk 1 above)
Comment (luitjens) [reply include/matx_tensor_ops.h]: fixed

File: include/matx_tensor_ops.h (hunk 3 above)
Comment (luitjens) [reply include/matx_tensor_ops.h:1176]: fixed

File: include/matx_tensor_ops.h (hunk 2 above)
Comment (luitjens) [reply include/matx_tensor_ops.h]: fixed

**Comment by cliffburdick** [comment]
Can you add to docs also?

**Comment by luitjens** [comment]
added to docs

**Review by cliffburdick — APPROVED** [review]

**Pull request merged**